File: problems/graphs/boggle_board/python/boggle_board.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False

    def search_in_children(self, letter: str):
        return self.children.get(letter, None)


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str):
        current_node: TrieNode = self.root
        for char in word:
            if char not in current_node.children:
                current_node.children[char] = TrieNode()
            current_node = current_node.children[char]
        current_node.is_end_of_word = True


def boggleBoard(board: list[list[str]], words: list[str]) -> dict:
    found_words: dict[str, list[tuple[int, int]]] = {}

    trie: Trie = Trie()
    for word in words:
        trie.insert(word)

    rows, cols = len(board), len(board[0])

    directions = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)]

    visited: list[list[bool]] = [[False for _ in range(cols)] for _ in range(rows)]

    def dfs(
        position: tuple[int, int],
        node: TrieNode,
        path: list[tuple[int, int]],
        current_word: str,
    ) -> None:
        x, y = position
        if not (0 <= x < rows and 0 <= y < cols) or visited[x][y]:
            return

        child_node = node.search_in_children(board[x][y])
        if not child_node:
            return

        visited[x][y] = True
        path.append((x, y))
        current_word += board[x][y]

        if child_node.is_end_of_word and current_word not in found_words:
            found_words[current_word] = path[:]

        for dx, dy in directions:
            dfs((x + dx, y + dy), child_node, path, current_word)

        visited[x][y] = False
        path.pop()

    for i in range(rows):
        for j in range(cols):
            dfs((i, j), trie.root, [], "")

    return {"words": found_words}

File: problems/graphs/boggle_board/js/boggle_board.js
const TrieNode = class {
  constructor() {
    this.children = {};
    this.isEndOfWord = false;
  }

  searchChild(character) {
    return this.children[character] || null;
  }
};

const Trie = class {
  constructor() {
    this.root = new TrieNode();
  }

  insert(word) {
    let currentNode = this.root;
    for (const char of word) {
      if (!currentNode.children[char])
        currentNode.children[char] = new TrieNode();
      currentNode = currentNode.children[char];
    }
    currentNode.isEndOfWord = true;
  }
};

export const boggleBoard = (board, words) => {
  const foundWords = [];
  const remainingWords = new Set(words);

  const trie = new Trie();
  for (const word of words) {
    trie.insert(word);
  }

  const rows = board.length;
  const cols = board[0].length;
  const directions = [
    [-1, -1],
    [-1, 0],
    [-1, 1],
    [0, -1],
    [0, 1],
    [1, -1],
    [1, 0],
    [1, 1],
  ];

  const visited = Array.from({ length: rows }, () => Array(cols).fill(false));

  const dfs = (position, node, path) => {
    const [x, y] = position;

    if (x < 0 || y < 0 || x >= rows || y >= cols || visited[x][y]) return;

    const childNode = node.searchChild(board[x][y]);
    if (!childNode) return;

    visited[x][y] = true;
    path.push(board[x][y]);

    if (childNode.isEndOfWord) {
      const word = path.join("");
      if (remainingWords.has(word)) {
        foundWords.push(word);
        remainingWords.delete(word);
        childNode.isEndOfWord = false;
      }
    }

    for (const [dx, dy] of directions) {
      dfs([x + dx, y + dy], childNode, path);
    }
    visited[x][y] = false;
    path.pop();
  };

  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[0].length; j++) {
      if (!remainingWords.size) break;
      dfs([i, j], trie.root, []);
    }
    if (!remainingWords.size) break;
  }
  return foundWords;
};
